# 1 - Data Preparation and Quality Control


## 1.1 Importing the Excel File
First, we imported the file into a Python environment to explore the dataset structure. To do this, we used the pandas library. The following code loaded the file and provided an initial overview of the data:

In [20]:
# Import required libraries
import pandas as pd

In [29]:
# Load dataset
try:
    # header=0 ensures that the first row is used as column headers.
    df = pd.read_excel("Data/AC2021_AnnualisedEntryExit.xlsx", header=0)
    print(f"Dataset loaded successfully with shape: {df.shape}")
except Exception as e:
    print(f"Error loading dataset from '{filepath}': {e}")
    raise

Dataset loaded successfully with shape: (458, 21)


In [23]:
# Show the columns
df.columns

Index(['Mode', 'NLC', 'ASC', 'Station', 'Coverage', 'Source',
       'Entries_Weekday', 'Entries_Friday', 'Entries_Saturday',
       'Entries_Sunday', 'Exits_Weekday', 'Exits_Friday', 'Exits_Saturday',
       'Exits_Sunday', 'Annualised_En_Ex', 'Total Entries per Station',
       'Total Exits per Station', 'Average Entries per Day ',
       '•\tAverage Exits per Day ', 'Total Weekly traffic', 'STDEV'],
      dtype='object')

In [24]:
# See the shape
df.shape

(458, 21)

In [25]:
# Displaying global information about the dataset
df.head()

,Mode,NLC,ASC,Station,Coverage,Source,Entries_Weekday,Entries_Friday,Entries_Saturday,Entries_Sunday,...,Exits_Friday,Exits_Saturday,Exits_Sunday,Annualised_En_Ex,Total Entries per Station,Total Exits per Station,Average Entries per Day,•\tAverage Exits per Day,Total Weekly traffic,STDEV
0,LU,625,KXXu,King's Cross St. Pancras,Station entry / exit,Gateline,76546.650000,86808.000000,89500.909091,69804.000000,...,90424.454545,89126.090909,71424.636364,3.673409e+07,322659.559091,328231.081818,80664.889773,82057.770455,650890.640909,8541.463211
1,LU,741,VICu,Victoria LU,Station entry / exit,Gateline,73590.850000,77402.181818,76219.250000,52525.666667,...,77246.545455,74958.000000,55658.444444,3.348093e+07,279737.948485,284217.989899,69934.487121,71054.497475,563955.938384,10232.456346
2,LU,635,LONu,London Bridge LU,Station entry / exit,Gateline,70445.550000,73371.200000,73043.363636,47535.272727,...,70908.400000,68767.727273,45522.454545,3.085708e+07,264395.386364,253551.981818,66098.846591,63387.995455,517947.368182,11393.242153
3,LU,669,OXCu,Oxford Circus,Station entry / exit,Gateline,69642.500000,68934.888889,68292.888889,43806.900000,...,83539.111111,90458.777778,49947.100000,3.286315e+07,250677.177778,301829.438889,62669.294444,75457.359722,552506.616667,15808.626048
4,LU,719,SFDu,Stratford,Station entry / exit,Gateline,62497.473684,67115.272727,71770.000000,48381.250000,...,68762.818182,72425.800000,48352.250000,2.910662e+07,249763.996411,253176.710287,62440.999103,63294.177572,502940.706699,9596.763929


These results give us an initial view of the content and diversity of the data. The dataset contains 458 records spread across 21 columns.

## 1.2 Checking for Null Values ​​and Duplicate Data
The quality of the analyses depends on the completeness and uniqueness of the data. We therefore checked for the presence of null values ​​and duplicates in the dataset:

### Checking for Missing Values
The following code counts the missing values ​​in each column:

In [26]:
missing_vals = df.isnull().sum()
print("Missing values for each column:")
for column, count in missing_vals.items():
    print(f"  {column}: {count}")

Missing values for each column:
  Mode: 0
  NLC: 0
  ASC: 0
  Station: 0
  Coverage: 0
  Source: 0
  Entries_Weekday: 0
  Entries_Friday: 0
  Entries_Saturday: 0
  Entries_Sunday: 0
  Exits_Weekday: 0
  Exits_Friday: 0
  Exits_Saturday: 0
  Exits_Sunday: 0
  Annualised_En_Ex: 0
  Total Entries per Station: 0
  Total Exits per Station: 0
  Average Entries per Day : 0
  •	Average Exits per Day : 0
  Total Weekly traffic: 0
  STDEV: 0


Conclusion: No records are incomplete, ensuring an analysis free from bias due to missing data.
### Duplicate Check
To ensure no data is duplicated, we performed:

In [27]:
num_duplicates = df.duplicated().sum()
print(f"Total duplicate rows found: {num_duplicates}")
if num_duplicates > 0:
    df = df.drop_duplicates()
    print("Duplicate rows have been removed.")
else:
    print("No duplicate rows found.")

Total duplicate rows found: 0
No duplicate rows found.


Conclusion: The dataset is free of duplicates, ensuring data quality for subsequent analyses.

### 1.3 Data Transformation and Typing
To ensure integrity during our analysis, we performed a thorough data typing check.

In [28]:
print("Data types for each column:")
data_types = df.dtypes
for column, dtype in data_types.items():
    print(f"  {column}: {dtype}")

Data types for each column:
  Mode: object
  NLC: int64
  ASC: object
  Station: object
  Coverage: object
  Source: object
  Entries_Weekday: float64
  Entries_Friday: float64
  Entries_Saturday: float64
  Entries_Sunday: float64
  Exits_Weekday: float64
  Exits_Friday: float64
  Exits_Saturday: float64
  Exits_Sunday: float64
  Annualised_En_Ex: float64
  Total Entries per Station: float64
  Total Exits per Station: float64
  Average Entries per Day : float64
  •	Average Exits per Day : float64
  Total Weekly traffic: float64
  STDEV: float64


### 1.3 Save the cleaned data.

In [ ]:
df.to_excel("cleaned_data.csv", index=False)

# Conclusion
The first phase of exploration confirmed the quality of the HSBC UK dataset. Key steps include:
## 2.1 Integrity Validation
No incomplete records were identified during the null value check. This ensures that the analysis is performed on a complete dataset, without bias induced by missing data.
## 2.2 Type Check and Correction
The check confirmed that numeric and categorical data were correctly defined.
## 2.3 Duplicate Analysis:
Running the *df.duplicated().sum()* function confirmed that no duplicates were present in the dataset. This ensures the reliability of aggregations and avoids bias during analysis.